# Progetto Foundations of Computer Science
## Anno accademico 2021-22
### Elisa Bernasconi 815551, Damiano Garofoli 837589, Susanna Maugeri 839365

You have to work on the files:
*  [Books](https://github.com/gdv/foundationsCS/raw/master/progetti/2021/Books.csv.gz)
*  [Book ratings](https://github.com/gdv/foundationsCS/raw/master/progetti/2021/Book-Ratings.csv.gz)
*  [Users](https://github.com/gdv/foundationsCS/raw/master/progetti/2021/Users.csv.gz)
*  [Goodbooks books](https://github.com/gdv/foundationsCS/raw/master/progetti/2021/goodbooks.csv.gz)
*  [Goodbooks ratings](https://github.com/gdv/foundationsCS/raw/master/progetti/2021/goodbooks-ratings.csv.gz)

### Notes

1.    It is mandatory to use GitHub for developing the project.
1.    The project must be a jupyter notebook.
1.    There is no restriction on the libraries that can be used, nor on the Python version.
1.    To read those files, you need to use the `encoding = 'latin-1'` option.
1.    All questions on the project **must** be asked in a public channel on [Zulip](https://focs.zulipchat.com), otherwise no  answer will be given.

### Importazione della libreria pandas e dei dataset necessari

In [1]:
import pandas as pd
import numpy as np

Iniziamo con l'importazione dei file .csv.

In [2]:
books = pd.read_csv("Data/Books.csv", encoding = 'latin-1', sep=';', low_memory=False, header=0)
books.tail(50)

ISBN                                         Book-Title  \
271309  0062554905  The color of light: Meditations for all of us ...   
271310  0304329045  How Can You Write a Poem When You Are Dying of...   
271311  0671015885               Star Trek: First Contact (Star Trek)   
271312  1555831141                                             Chrome   
271313  1563332442  Two Thousand Sixty-Nine: A Science Fiction Tri...   
271314  1902852036                                All the Queen's Men   
271315  0441011799                                    Singularity Sky   
271316  359610601X  Der gelbe Stern: Die Judenvernichtung in Europ...   
271317  0375415343                                  A Whistling Woman   
271318  3404204433                           Am Ende des Regenbogens.   
271319  3462017942              Die Massenpsychologie des Faschismus.   
271320  3453047192  Die amerikanische Zumutung: Plädoyers gegen...   
271321  9813056398      Broken Mirror : True Stories About Drug Abuse   
271322  000637610X                                   You Got an Ology   
271323  0878576568              Biopotency: A Guide in Sexual Success   
271324  0747204020                  Illustrated Encyclopedia of Cacti   
271325  0070542023                       Basic Scientific Subroutines   
271326  0070704309                  Electronic filter design handbook   
271327  0471915645  Introductory Digital Signal Processing with Co...   
271328  8420614556  Lewis Carroll: A Traves Del Espejo Y Lo Que Al...   
271329  3320016822  Urteil ohne Prozess: Margot Honecker gegen Oss...   
271330  3423200944  Hokuspokus, liebe mich. Und f�¼nfzig ande...   
271331  3453065123  Ich war Hitlerjunge Salomon. ( erlebt und erfa...   
271332  3525335423                Das Deutsche Kaiserreich 1871-1918.   
271333  3548740146  Wicca. Eine Einf�¼hrung in wei��e Magie.   
271334  381440176X  Ein Fall fÃ?¼r TKKG, Bd.50, Sklaven fÃ?Â...   
271335  3893312307  Die Vereinten Nationen: Zwischen Anspruch und ...   
271336  0971854823  The Female Power Within: A Guide to Living a G...   
271337  0316640786  Christie's Collectables: Blue and White China ...   
271338  3257217323                       Schmatz. Oder Die Sackgasse.   
271339  3596156904                                              Amok.   
271340  1874166633  Introducing Nietzsche (Foundations in Children...   
271341  0130897930                 Core Web Programming (2nd Edition)   
271342  020130998X  The Unified Modeling Language Reference Manual...   
271343  2268032019         Petite histoire de la d�©sinformation   
271344  0684860112        Driving to Detroit: Memoirs of a Fast Woman   
271345  0395264707                                         Dreamsnake   
271346  3442150663                                        Der Mossad.   
271347  0231128444                      Slow Food(The Case For Taste)   
271348  0520242335  Strong Democracy : Participatory Politics for ...   
271349  0762412119  Burpee Gardening Cyclopedia: A Concise, Up to ...   
271350  1582380805  Tropical Rainforests: 230 Species in Full Colo...   
271351  1845170423                                  Cocktail Classics   
271352  014002803X                                  Anti Death League   
271353  0449906736      Flashpoints: Promise and Peril in a New World   
271354  0440400988                         There's a Bat in Bunk Five   
271355  0525447644                            From One to One Hundred   
271356  006008667X  Lily Dale : The True Story of the Town that Ta...   
271357  0192126040                        Republic (World's Classics)   
271358  0767409752  A Guided Tour of Rene Descartes' Meditations o...   

                    Book-Author Year-Of-Publication  \
271309          Perry Tilleraas                1988   
271310              John Harold                1994   
271311             J.M. Dillard                1997   
271312             George Nadar                1987   
271313         

In [3]:
book_rat = pd.read_csv("Data/Book-Ratings.csv", encoding = 'latin-1', sep=';', low_memory=False, header=0)
book_rat.head()

User-ID        ISBN  Book-Rating
0   276725  034545104X            0
1   276726  0155061224            5
2   276727  0446520802            0
3   276729  052165615X            3
4   276729  0521795028            6

In [4]:
users = pd.read_csv("Data/Users.csv", encoding = 'latin-1', sep=';', low_memory=False, header=0)
users

User-ID                            Location   Age
0             1                  nyc, new york, usa   NaN
1             2           stockton, california, usa  18.0
2             3     moscow, yukon territory, russia   NaN
3             4           porto, v.n.gaia, portugal  17.0
4             5  farnborough, hants, united kingdom   NaN
...         ...                                 ...   ...
278853   278854               portland, oregon, usa   NaN
278854   278855  tacoma, washington, united kingdom  50.0
278855   278856           brampton, ontario, canada   NaN
278856   278857           knoxville, tennessee, usa   NaN
278857   278858                dublin, n/a, ireland   NaN

[278858 rows x 3 columns]

In [5]:
good_book = pd.read_csv("Data/goodbooks.csv", sep=',', encoding = 'latin-1', low_memory=False, header=0)
good_book.tail(50)

book_id  goodreads_book_id  best_book_id   work_id  books_count  \
49       50              30119         30119     30518           45   
50       51             256683        256683   2267189          178   
51       52             428263        428263   2675454          185   
52       53             113436        113436   3178011          217   
53       54                 11        386162   3078186          257   
54       55               5129          5129   3204877          515   
55       56            1162543       1162543   2960529          183   
56       57              37435         37435   3275013          125   
57       58               2956          2956   1835605         2277   
58       59              24178         24178    987048          180   
59       60               1618          1618   4259809          183   
60       61           22557272      22557272  41107568           14   
61       62             119322        119322   1536771          287   
62       63               6185          6185   1565818         2498   
63       64              10917         10917   1639903          125   
64       65               4981          4981   1683562          241   
65       66              18405         18405   3358283          409   
66       67             128029        128029   3271379          197   
67       68              22628         22628   2236198          128   
68       69           11735983      11735983  15524542          164   
69       70             375802        375802   2422333          224   
70       71              18490         18490   4836639         2618   
71       72              11588         11588    849585          289   
72       73            1656001       1656001   3328799          161   
73       74              99561         99561    919292          187   
74       75             227443        227443   3185154          193   
75       76              14935         14935   2809709         1969   
76       77              38709         38709   1679789          138   
77       78               5139          5139     38765          129   
78       79               1381          1381   3356006         1703   
79       80             157993        157993   2180358         1708   
80       81               7445          7445   2944133           92   
81       82               1845          1845   3284484          108   
82       83               1953          1953   2956372          525   
83       84               7677       6424171   3376836          176   
84       85             370493        370493     30530           81   
85       86              32542         32542   1804929          158   
86       87               1617          1617    265616          109   
87       88            6442769       6442769   3364505          178   
88       89              21787         21787    992628          129   
89       90             231804        231804   1426690          156   
90       91            6186357       6186357   6366642          147   
91       92               1202          1202      5397          124   
92       93               2998          2998   3186437         1350   
93       94                320           320   3295655          555   
94       95               5297          5297   1858012         2303   
95       96           13536860      13536860  18034963          133   
96       97              17245         17245   3165724         2207   
97       98            5060378       5060378   6976108          209   
98       99           11857408      11857408  16813814          147   

          isbn        isbn13  \
49    60513039  9.780061e+12   
50  1416914285  9.781417e+12   
51   316160202  9.780316e+12   
52   375826696  9.780376e+12   
53   345391802  9.780345e+12   
54    60929871  9.780061e+12   
55  031606792X  9.780316e+12   
56   142001740  9.780142e+12   
57   142437174  9.780142e+12   
58    64410935  9.780064e+12   
59  1400032717  9.781400e+12  

In [6]:
good_rat = pd.read_csv("Data/goodbooks-ratings.csv", encoding = 'latin-1', sep=',', low_memory=False, header=0)
good_rat.head()

user_id  book_id  rating
0        1      258       5
1        2     4081       4
2        2      260       5
3        2     9296       5
4        2     2318       3

L'importazione dei file .csv è andata a buon fine, tuttavia nonostante l'utilizzo il tipo di encoding _Latin-1_ notiamo che vi sono dei caratteri che non sono stati decodificati nel modo corretto.

### 1. Normalize the location field of *Users* dataset, splitting into city, region, country.

Verifichiamo di che tipo sia ogni valore:

In [7]:
type(users['Location'][0])

str

Si tratta di stringhe, perciò per dividere i campi si può utilizzare uno split.

Verifichiamo che per tutte le righe vi siano le tre informazioni necessarie:

In [8]:
cont=np.zeros(10,dtype=int)

for i, row in users.iterrows():
    lung = len(row['Location'].split(','))
    if lung == 1:
        cont[0]+=1
    elif lung == 2:
        cont[1]+=1
    elif lung == 3:
        cont[2]+=1
    elif lung == 4:
        cont[3]+=1
    elif lung == 5:
        cont[4]+=1
    elif lung == 6:
        cont[5]+=1
    elif lung == 7:
        cont[6]+=1
    elif lung == 8:
        cont[7]+=1
    elif lung == 9:
        cont[8]+=1
    else:
        cont[9]+=1
        
for i in range(0,10):
    if i==0:
        print('Righe con', i+1, 'elemento in Location:', cont[i])
    elif i==9:
         print('Righe con 0,10 o più elementi in Location:', cont[i])
    else:
        print('Righe con', i+1, 'elementi in Location:', cont[i])

Righe con 1 elemento in Location: 1
Righe con 2 elementi in Location: 1
Righe con 3 elementi in Location: 277348
Righe con 4 elementi in Location: 1417
Righe con 5 elementi in Location: 72
Righe con 6 elementi in Location: 11
Righe con 7 elementi in Location: 4
Righe con 8 elementi in Location: 2
Righe con 9 elementi in Location: 2
Righe con 0,10 o più elementi in Location: 0


Il totale delle righe che non hanno l'esatto numero di elementi che ci servono sono

In [9]:
sbagliati = sum(cont)-cont[2]
sbagliati

1510

Il numero totale di righe della tabella _users_ è

In [10]:
totale = users.shape[0]
totale

278858

Perciò la percentuale di righe con informazioni mancanti o inesatte è del

In [11]:
print(round(sbagliati/totale, 3),'%')

0.005 %


Osserviamo nello specifico alcune di queste osservazioni:

In [12]:
print(users.loc[6117]['Location'])
print(users.loc[12512]['Location'])
print(users.loc[6457]['Location'])
print(users.loc[3924]['Location'])

birmingham, alabama, alabama, usa
nanaimo, bc, v9v 1h4, british columbia, canada
mayfield, sussex (haha, you listed stroud!), united kingdom
bombay, maharashtra, heard of it? :o), india


Alcune presentano valori ripetuti, senza significato o addirittura commenti personali.

Decidiamo pertanto di eliminare tali osservazioni dal dataset, poiché ne rappresentano una piccola percentuale e poiché potrebbero essere portatrici di informazioni errate o superflue. 

In [13]:
indici = []
for i, row in users.iterrows():
    lung = len(row['Location'].split(','))
    if lung != 3:
        indici.append(i)
        
users = users.drop(indici, axis=0)
len(users)

277348

Dunque il numero di osservazioni che consideriamo è 277348 e con queste possiamo ricavare le tre colonne richieste: _city, region_ e _country_.

In [14]:
users['city'] = users['Location'].str.split(expand=True, pat=',')[0]
users['region'] = users['Location'].str.split(expand=True, pat=',')[1]
users['country'] = users['Location'].str.split(expand=True, pat=',')[2]

In [15]:
users.head()

User-ID                            Location   Age         city  \
0        1                  nyc, new york, usa   NaN          nyc   
1        2           stockton, california, usa  18.0     stockton   
2        3     moscow, yukon territory, russia   NaN       moscow   
3        4           porto, v.n.gaia, portugal  17.0        porto   
4        5  farnborough, hants, united kingdom   NaN  farnborough   

             region          country  
0          new york              usa  
1        california              usa  
2   yukon territory           russia  
3          v.n.gaia         portugal  
4             hants   united kingdom

L'operazione è andata a buon fine.

### 2. For each book in the *Books* dataset, compute its average rating.

In [16]:
books.columns

Index(['ISBN', 'Book-Title', 'Book-Author', 'Year-Of-Publication', 'Publisher',
       'Image-URL-S', 'Image-URL-M', 'Image-URL-L'],
      dtype='object')

In [17]:
book_rat.columns

Index(['User-ID', 'ISBN', 'Book-Rating'], dtype='object')

Calcoliamo l'_average rating_ per ciascun libro della tabella _book_rat_ e poi operiamo un merge del risultato con la tabella _books_.

In [18]:
ave_rat = book_rat.groupby(['ISBN'])['Book-Rating'].mean()
ave_rat.head()

ISBN
 0330299891    3.0
 0375404120    1.5
 0586045007    0.0
 9022906116    3.5
 9032803328    0.0
Name: Book-Rating, dtype: float64

In [19]:
books= pd.merge(books, ave_rat, on='ISBN')
books.rename({'Book-Rating': 'Average-Rating'}, axis='columns', inplace=True)
books[['ISBN', 'Book-Title', 'Average-Rating']].head()

ISBN                                         Book-Title  \
0  0195153448                                Classical Mythology   
1  0002005018                                       Clara Callan   
2  0060973129                               Decision in Normandy   
3  0374157065  Flu: The Story of the Great Influenza Pandemic...   
4  0393045218                             The Mummies of Urumchi   

   Average-Rating  
0        0.000000  
1        4.928571  
2        5.000000  
3        4.272727  
4        0.000000

### 3. For each book in the *GoodBooks* dataset, compute its average rating.

Osserviamo i nomi di tutte le colonne del dataframe in questione:

In [20]:
good_book.columns

Index(['book_id', 'goodreads_book_id', 'best_book_id', 'work_id',
       'books_count', 'isbn', 'isbn13', 'authors', 'original_publication_year',
       'original_title', 'title', 'language_code', 'average_rating',
       'ratings_count', 'work_ratings_count', 'work_text_reviews_count',
       'ratings_1', 'ratings_2', 'ratings_3', 'ratings_4', 'ratings_5',
       'image_url', 'small_image_url'],
      dtype='object')

Notiamo che vi è già una colonna _average_rating_, tuttavia proviamo a calcolare la media per verificare se i valori coincidono.

-- Elisa : Consideriamo ad esempio gli elementi della riga con indice 0 --

Sommiamo i valori delle colonne _ratings_1, ratings_2, ratings_3, ratings_4_ e _ratings_5_ per capire se come denominatore per la media debba essere utilizzato l'attributo _ratings_count_ oppure _work_ratings_count_.

In [21]:
print('Somma dei ratings: ', good_book['ratings_1'][0] + good_book['ratings_2'][0] + good_book['ratings_3'][0] + good_book['ratings_4'][0] + good_book['ratings_5'][0]) 
print('Ratings count: ', good_book['ratings_count'][0])
print('Work ratings count: ', good_book['work_ratings_count'][0])

Somma dei ratings:  4942365
Ratings count:  4780653
Work ratings count:  4942365


Capiamo che la somma delle 5 colonne di ratings è contenuta in _work_ratings_count_.

Calcoliamo dunque la media:

In [22]:
good_book['media'] = (good_book['ratings_1'] + good_book['ratings_2']*2 + good_book['ratings_3']*3 + good_book['ratings_4']*4 +good_book['ratings_5']*5)/good_book['work_ratings_count']
good_book['media'] = round(good_book['media'], ndigits=2)

In [23]:
good_book[['isbn', 'original_title', 'media', 'average_rating']]

isbn                            original_title  media  \
0    439023483                          The Hunger Games   4.34   
1    439554934  Harry Potter and the Philosopher's Stone   4.44   
2    316015849                                  Twilight   3.57   
3     61120081                     To Kill a Mockingbird   4.25   
4    743273567                          The Great Gatsby   3.89   
..         ...                                       ...    ...   
94   375751513                The Picture of Dorian Gray   4.06   
95   345803507                        Fifty Shades Freed   3.88   
96   393970124                                   Dracula   3.98   
97   307269981               Flickan som lekte med elden   4.22   
98  1612130585                       Fifty Shades Darker   3.87   

    average_rating  
0             4.34  
1             4.44  
2             3.57  
3             4.25  
4             3.89  
..             ...  
94            4.06  
95            3.88  
96            3.98  
97            4.22  
98            3.87  

[99 rows x 4 columns]

Effettivamente i valori del nostro calcolo, _media,_ e della colonna _average_rating_ coincidono.

### 4. Merge together all rows sharing the same book title, author and publisher. We will call the resulting datset `merged books`. The books that have not been merged together will not appear in `merged books`.

Con l'opzione _count_ otteniamo l'informazione circa quante osservazioni siamo state accorpate in ciascun gruppo di chiave _(Book-Title, Book-Author, Publisher)_.

In [24]:
merged_books = books.groupby(by=["Book-Title", "Book-Author", "Publisher"], as_index=False).count()
print(len(merged_books))
merged_books.head()

264961


Book-Title       Book-Author  \
0   A Light in the Storm: The Civil War Diary of ...       Karen Hesse   
1                              Always Have Popsicles    Rebecca Harvin   
2               Apple Magic (The Collector's series)  Martina Boudreau   
3   Ask Lily (Young Women of Faith: Lily Series, ...      Nancy N. Rue   
4   Beyond IBM: Leadership Marketing and Finance ...        Lou Mobley   

                     Publisher  ISBN  Year-Of-Publication  Image-URL-S  \
0  Hyperion Books for Children     1                    1            1   
1            Rebecca L. Harvin     1                    1            1   
2           Amer Cooking Guild     1                    1            1   
3                   Zonderkidz     1                    1            1   
4       Teleonet, Incorporated     1                    1            1   

   Image-URL-M  Image-URL-L  Average-Rating  
0            1            1               1  
1            1            1               1  
2            1            1               1  
3            1            1               1  
4            1            1               1

In [25]:
indici = merged_books[merged_books['ISBN'] == 1 ].index
merged_books.drop(indici, inplace=True)
print(len(merged_books))
merged_books.head()

4686


Book-Title  \
41                           !%@ (A Nutshell handbook)   
58   'A Hell of a Place to Lose a Cow': An American...   
252  10,000 dreams interpreted: A dictionary of dreams   
461                                   101 Famous Poems   
709                 15 Houseplants Even You Can't Kill   

                 Book-Author                       Publisher  ISBN  \
41             Donnalyn Frey                        O'Reilly     2   
58               Tim Brookes             National Geographic     2   
252  Gustavus Hindman Miller       Barnes &amp; Nobles Books     2   
461              Roy J. Cook  McGraw-Hill/Contemporary Books     3   
709                Joe Elder               Berkley Pub Group     2   

     Year-Of-Publication  Image-URL-S  Image-URL-M  Image-URL-L  \
41                     2            2            2            2   
58                     2            2            2            2   
252                    2            2            2            2   
461                    3            3            3            3   
709                    2            2            2            2   

     Average-Rating  
41                2  
58                2  
252               2  
461               3  
709               2

Così abbiamo ottenuto un dataframe contenente solo i 4686 libri che hanno avuto due o più edizioni.

Infine, eliminiamo le colonne _ISBN, Year-Of-Publication, Image-URL-S, Image-URL-M, Image-URL-L, Average-Rating,_ poiché esse non risultano essere più informative.

In [26]:
merged_books = merged_books.drop(['ISBN', 'Year-Of-Publication', 'Image-URL-S', 'Image-URL-M', 'Image-URL-L', 'Average-Rating'], axis=1)
merged_books.head()

Book-Title  \
41                           !%@ (A Nutshell handbook)   
58   'A Hell of a Place to Lose a Cow': An American...   
252  10,000 dreams interpreted: A dictionary of dreams   
461                                   101 Famous Poems   
709                 15 Houseplants Even You Can't Kill   

                 Book-Author                       Publisher  
41             Donnalyn Frey                        O'Reilly  
58               Tim Brookes             National Geographic  
252  Gustavus Hindman Miller       Barnes &amp; Nobles Books  
461              Roy J. Cook  McGraw-Hill/Contemporary Books  
709                Joe Elder               Berkley Pub Group

### 5. For each book in `merged books` compute its average rating.

The average is computed considering all books in `books` that have been merged.

In [27]:
books.columns

Index(['ISBN', 'Book-Title', 'Book-Author', 'Year-Of-Publication', 'Publisher',
       'Image-URL-S', 'Image-URL-M', 'Image-URL-L', 'Average-Rating'],
      dtype='object')

In [28]:
merged_books.columns

Index(['Book-Title', 'Book-Author', 'Publisher'], dtype='object')

Facciamo un _left join_ tra le tabelle _merged_books_ e _books,_ in modo tale che vengano tenuti tutti i libri che hanno avuto almeno due edizioni, con le informazioni relative a ciascuna edizione in righe diverse del dataframe.

In [29]:
merged_books2 = merged_books.merge(books, on=['Book-Title', 'Book-Author', 'Publisher'], how='left')
merged_books2.head()

Book-Title              Book-Author  \
0                          !%@ (A Nutshell handbook)            Donnalyn Frey   
1                          !%@ (A Nutshell handbook)            Donnalyn Frey   
2  'A Hell of a Place to Lose a Cow': An American...              Tim Brookes   
3  'A Hell of a Place to Lose a Cow': An American...              Tim Brookes   
4  10,000 dreams interpreted: A dictionary of dreams  Gustavus Hindman Miller   

                   Publisher        ISBN Year-Of-Publication  \
0                   O'Reilly  1565920465                1994   
1                   O'Reilly  1565920317                1993   
2        National Geographic  0792276833                2000   
3        National Geographic  0792277295                2001   
4  Barnes &amp; Nobles Books  1566196256                1995   

                                         Image-URL-S  \
0  http://images.amazon.com/images/P/1565920465.0...   
1  http://images.amazon.com/images/P/1565920317.0...   
2  http://images.amazon.com/images/P/0792276833.0...   
3  http://images.amazon.com/images/P/0792277295.0...   
4  http://images.amazon.com/images/P/1566196256.0...   

                                         Image-URL-M  \
0  http://images.amazon.com/images/P/1565920465.0...   
1  http://images.amazon.com/images/P/1565920317.0...   
2  http://images.amazon.com/images/P/0792276833.0...   
3  http://images.amazon.com/images/P/0792277295.0...   
4  http://images.amazon.com/images/P/1566196256.0...   

                                         Image-URL-L  Average-Rating  
0  http://images.amazon.com/images/P/1565920465.0...            0.00  
1  http://images.amazon.com/images/P/1565920317.0...            6.00  
2  http://images.amazon.com/images/P/0792276833.0...            0.00  
3  http://images.amazon.com/images/P/0792277295.0...            3.40  
4  http://images.amazon.com/images/P/1566196256.0...            7.25

Infine, possiamo calcolare l'_Average-Rating_ per le edizioni di ciascun libro.

In [30]:
ave_rat = merged_books2.groupby(['Book-Title', 'Book-Author', 'Publisher'])['Average-Rating'].mean()
merged_books = pd.merge(merged_books, ave_rat, on=['Book-Title', 'Book-Author', 'Publisher'])
merged_books.head()

Book-Title              Book-Author  \
0                          !%@ (A Nutshell handbook)            Donnalyn Frey   
1  'A Hell of a Place to Lose a Cow': An American...              Tim Brookes   
2  10,000 dreams interpreted: A dictionary of dreams  Gustavus Hindman Miller   
3                                   101 Famous Poems              Roy J. Cook   
4                 15 Houseplants Even You Can't Kill                Joe Elder   

                        Publisher  Average-Rating  
0                        O'Reilly        3.000000  
1             National Geographic        1.700000  
2       Barnes &amp; Nobles Books        6.958333  
3  McGraw-Hill/Contemporary Books        3.111111  
4               Berkley Pub Group        0.000000

### 6. For each book in `merged books` compute the minimum and maximum of the average ratings over all corresponding books in the `books` dataset.

Hence for each book in `merged books` we will have exactly two values (a minimum and a maximum)

Utilizziamo la tabella _merged_books2_ che contiene le diverse edizioni di ciascun libro di _merged_books._

In [31]:
minimi = merged_books2.groupby(['Book-Title', 'Book-Author', 'Publisher'], as_index=False)['Average-Rating'].min()
massimi = merged_books2.groupby(['Book-Title', 'Book-Author', 'Publisher'], as_index=False)['Average-Rating'].max()

minimi.rename({'Average-Rating': 'Minimum-Average-Rating'}, axis='columns', inplace=True)
massimi.rename({'Average-Rating': 'Maximum-Average-Rating'}, axis='columns', inplace=True)
massimi.head()

Book-Title              Book-Author  \
0                          !%@ (A Nutshell handbook)            Donnalyn Frey   
1  'A Hell of a Place to Lose a Cow': An American...              Tim Brookes   
2  10,000 dreams interpreted: A dictionary of dreams  Gustavus Hindman Miller   
3                                   101 Famous Poems              Roy J. Cook   
4                 15 Houseplants Even You Can't Kill                Joe Elder   

                        Publisher  Maximum-Average-Rating  
0                        O'Reilly                    6.00  
1             National Geographic                    3.40  
2       Barnes &amp; Nobles Books                    7.25  
3  McGraw-Hill/Contemporary Books                    5.00  
4               Berkley Pub Group                    0.00

In [32]:
merged_books = pd.merge(merged_books, minimi, on=['Book-Title', 'Book-Author', 'Publisher'])
merged_books = pd.merge(merged_books, massimi, on=['Book-Title', 'Book-Author', 'Publisher'])
merged_books.head()

Book-Title              Book-Author  \
0                          !%@ (A Nutshell handbook)            Donnalyn Frey   
1  'A Hell of a Place to Lose a Cow': An American...              Tim Brookes   
2  10,000 dreams interpreted: A dictionary of dreams  Gustavus Hindman Miller   
3                                   101 Famous Poems              Roy J. Cook   
4                 15 Houseplants Even You Can't Kill                Joe Elder   

                        Publisher  Average-Rating  Minimum-Average-Rating  \
0                        O'Reilly        3.000000                0.000000   
1             National Geographic        1.700000                0.000000   
2       Barnes &amp; Nobles Books        6.958333                6.666667   
3  McGraw-Hill/Contemporary Books        3.111111                0.000000   
4               Berkley Pub Group        0.000000                0.000000   

   Maximum-Average-Rating  
0                    6.00  
1                    3.40  
2                    7.25  
3                    5.00  
4                    0.00

### 7. For each book in `goodbooks`, compute the list of its authors. Assuming that the number of reviews with a text (column `work_text_reviews_count`) is split equally among all authors, find for each authors the total number of reviews with a text. We will call this quantity the *shared number of reviews with a text*.

In [33]:
good_book['List-Of-Authors'] = good_book['authors'].str.split(',')
good_book[['book_id', 'original_title', 'authors', 'List-Of-Authors']].head()

book_id                            original_title  \
0        1                          The Hunger Games   
1        2  Harry Potter and the Philosopher's Stone   
2        3                                  Twilight   
3        4                     To Kill a Mockingbird   
4        5                          The Great Gatsby   

                        authors                  List-Of-Authors  
0               Suzanne Collins                [Suzanne Collins]  
1  J.K. Rowling, Mary GrandPré  [J.K. Rowling,  Mary GrandPré]  
2               Stephenie Meyer                [Stephenie Meyer]  
3                    Harper Lee                     [Harper Lee]  
4           F. Scott Fitzgerald            [F. Scott Fitzgerald]

In [34]:
import math

In [35]:
from gensim.parsing.preprocessing import strip_multiple_whitespaces

ModuleNotFoundError: No module named 'gensim'

In [36]:
autori = dict()
for ind, row in good_book.iterrows():
    lista = row['authors'].split(',')    #lista con 1 o più nomi
    for nome in lista:
        nome = nome.strip()
        #nome = strip_multiple_whitespaces(nome)
        if nome not in autori:
            autori[nome] = round(row['work_text_reviews_count']/len(lista), 3)
        else:
            autori[nome] += row['work_text_reviews_count']/len(lista)

In [37]:
shared_number_of_reviews_with_a_text_GB = autori
shared_number_of_reviews_with_a_text_GB

{'Suzanne Collins': 340066.0,
 'J.K. Rowling': 136668.0,
 'Mary GrandPré': 136668.0,
 'Stephenie Meyer': 258573.0,
 'Harper Lee': 72586.0,
 'F. Scott Fitzgerald': 51992.0,
 'John Green': 230584.0,
 'J.R.R. Tolkien': 52986.0,
 'J.D. Salinger': 44920.0,
 'Dan Brown': 66672.0,
 'Jane Austen': 52994.0,
 'Khaled Hosseini': 103375.0,
 'Veronica Roth': 156896.0,
 'George Orwell': 50644.667,
 'Erich Fromm': 15172.667,
 'Celâl Ã\x9cster': 15172.667,
 'Anne Frank': 6941.667,
 'Eleanor Roosevelt': 6941.667,
 'B.M. Mooyaart-Doubleday': 6941.667,
 'Stieg Larsson': 46983.0,
 'Reg Keeland': 46983.0,
 'Rufus Beck': 12033.0,
 'Alice Sebold': 36642.0,
 'William Golding': 26886.0,
 'William Shakespeare': 7389.0,
 'Robert           Jackson': 7389.0,
 'Gillian Flynn': 121614.0,
 'Kathryn Stockett': 78204.0,
 'John Steinbeck': 24642.0,
 'Arthur Golden': 25605.0,
 'E.L. James': 128776.0,
 'Paulo Coelho': 27890.5,
 'Alan R. Clarke': 27890.5,
 'Lois Lowry': 54084.0,
 'C.S. Lewis': 15186.0,
 'Audrey Niffenegg

In [87]:
shared=pd.DataFrame(shared_number_of_reviews_with_a_text_GB.items())
shared.rename({0: "Author",1:"reviews"},axis='columns',inplace=True)
shared

Author     reviews
0      Suzanne Collins  340066.000
1         J.K. Rowling  136668.000
2       Mary GrandPré  136668.000
3      Stephenie Meyer  258573.000
4           Harper Lee   72586.000
..                 ...         ...
104        Oscar Wilde    9823.500
105  Jeffrey Eugenides    9823.500
106        Bram Stoker    5754.333
107      Nina Auerbach    5754.333
108      David J. Skal    5754.333

[109 rows x 2 columns]

### 8. For each year of publication, determine the author that has the largest value of the shared number of reviews with a text.

In [38]:
good_book.columns

Index(['book_id', 'goodreads_book_id', 'best_book_id', 'work_id',
       'books_count', 'isbn', 'isbn13', 'authors', 'original_publication_year',
       'original_title', 'title', 'language_code', 'average_rating',
       'ratings_count', 'work_ratings_count', 'work_text_reviews_count',
       'ratings_1', 'ratings_2', 'ratings_3', 'ratings_4', 'ratings_5',
       'image_url', 'small_image_url', 'media', 'List-Of-Authors'],
      dtype='object')

Creiamo un dizionario in cui le chiavi sono gli anni e i valori sono le liste di autori che hanno pubblicato uno o più libri in quell'anno.

In [39]:
anni_autori = dict()
for i, row in good_book.iterrows():
    anno = int(row['original_publication_year'])
    if anno not in anni_autori:
        anni_autori[anno] = row['authors'].strip().split(',')
       # anni_autori[anno] = strip_multiple_whitespaces(row['authors'].strip()).split(',')
    else:
        anni_autori[anno].extend(row['authors'].strip().split(','))
        #anni_autori[anno].extend(strip_multiple_whitespaces(row['authors'].strip()).split(','))
        
anni_autori

{2008: ['Suzanne Collins', 'Stephenie Meyer', 'Stephenie Meyer', 'John Green'],
 1997: ['J.K. Rowling', ' Mary GrandPré', 'Arthur Golden'],
 2005: ['Stephenie Meyer',
  'Stieg Larsson',
  ' Reg Keeland',
  'J.K. Rowling',
  ' Mary GrandPré',
  'Rick Riordan',
  'Markus Zusak',
  'John Green',
  'Jeannette Walls',
  'Steven D. Levitt',
  ' Stephen J. Dubner'],
 1960: ['Harper Lee'],
 1925: ['F. Scott Fitzgerald'],
 2012: ['John Green', 'Gillian Flynn', 'Veronica Roth', 'E.L. James'],
 1937: ['J.R.R. Tolkien', 'John Steinbeck'],
 1951: ['J.D. Salinger'],
 2000: ['Dan Brown', 'J.K. Rowling', ' Mary GrandPré'],
 1813: ['Jane Austen'],
 2003: ['Khaled Hosseini',
  'J.K. Rowling',
  ' Mary GrandPré',
  'Dan Brown',
  'Audrey Niffenegger',
  'Mark Haddon',
  'Lauren Weisberger'],
 2011: ['Veronica Roth', 'E.L. James', 'E.L. James'],
 1949: ['George Orwell', ' Erich Fromm', ' Celâl Ã\x9cster'],
 1945: ['George Orwell'],
 1947: ['Anne Frank', ' Eleanor Roosevelt', ' B.M. Mooyaart-Doubleday

Da ogni lista eliminiamo i duplicati rendendola prima _set_ e poi nuovamente lista:

In [40]:
for key, value in anni_autori.items():
    anni_autori[key] = list(set(anni_autori[key]))
anni_autori

{2008: ['John Green', 'Suzanne Collins', 'Stephenie Meyer'],
 1997: [' Mary GrandPré', 'J.K. Rowling', 'Arthur Golden'],
 2005: [' Reg Keeland',
  'Stephenie Meyer',
  'John Green',
  'Rick Riordan',
  ' Stephen J. Dubner',
  'J.K. Rowling',
  'Jeannette Walls',
  'Steven D. Levitt',
  ' Mary GrandPré',
  'Markus Zusak',
  'Stieg Larsson'],
 1960: ['Harper Lee'],
 1925: ['F. Scott Fitzgerald'],
 2012: ['John Green', 'Veronica Roth', 'E.L. James', 'Gillian Flynn'],
 1937: ['John Steinbeck', 'J.R.R. Tolkien'],
 1951: ['J.D. Salinger'],
 2000: ['Dan Brown', ' Mary GrandPré', 'J.K. Rowling'],
 1813: ['Jane Austen'],
 2003: ['Khaled Hosseini',
  'Mark Haddon',
  'J.K. Rowling',
  'Lauren Weisberger',
  ' Mary GrandPré',
  'Dan Brown',
  'Audrey Niffenegger'],
 2011: ['E.L. James', 'Veronica Roth'],
 1949: [' Erich Fromm', ' Celâl Ã\x9cster', 'George Orwell'],
 1945: ['George Orwell'],
 1947: [' B.M. Mooyaart-Doubleday', ' Eleanor Roosevelt', 'Anne Frank'],
 2009: ['James Dashner', 'Suz

Osserviamo le chiavi del dizionario, che dovrebbero essere gli anni:

In [41]:
anni_autori.keys()

dict_keys([2008, 1997, 2005, 1960, 1925, 2012, 1937, 1951, 2000, 1813, 2003, 2011, 1949, 1945, 1947, 2009, 1999, 1954, 2010, 2002, 1998, 2007, 1595, 1988, 1993, 1950, 1996, 2006, 1868, 1847, 2001, 1953, 1974, 1979, 1932, 1884, 1952, 2015, 1995, 2004, 1969, 1936, 1985, 1818, 1977, 1811, -720, 1946, 1859, 1990, 1964, 1989, 1958, 1973, 1967, 1911, 1891, 1897])

Vi è un valore anomalo: -720. Osserviamo il titolo a cui è associato questo anno di pubblicazione.

In [42]:
good_book[good_book['original_publication_year']== -720]['original_title']

78    Ὀδύσσεια
Name: original_title, dtype: object

Decidiamo di eliminarlo dal dizionario:

In [43]:
del anni_autori[-720]
anni_autori.keys()

dict_keys([2008, 1997, 2005, 1960, 1925, 2012, 1937, 1951, 2000, 1813, 2003, 2011, 1949, 1945, 1947, 2009, 1999, 1954, 2010, 2002, 1998, 2007, 1595, 1988, 1993, 1950, 1996, 2006, 1868, 1847, 2001, 1953, 1974, 1979, 1932, 1884, 1952, 2015, 1995, 2004, 1969, 1936, 1985, 1818, 1977, 1811, 1946, 1859, 1990, 1964, 1989, 1958, 1973, 1967, 1911, 1891, 1897])

Ora non notiamo valori anomali.

Per ciascun anno estraiamo l'autore che ha il numero massimo di _shared_reviews_with_a_text._

In [108]:
anno_aut_max = dict()
for anno in anni_autori.keys():
    autori=anni_autori[anno]
    max=0
    vicitore=''
    for autore in autori:
        autore=autore.strip()
        #autore=strip_multiple_whitespaces(autore).strip()
        if shared_number_of_reviews_with_a_text_GB[autore]>max:
            max=shared_number_of_reviews_with_a_text_GB[autore]
            vincitore=autore
    anno_aut_max[anno]=vincitore
anno_aut_max

{2008: 'Suzanne Collins',
 1997: 'Mary GrandPré',
 2005: 'Stephenie Meyer',
 1960: 'Harper Lee',
 1925: 'F. Scott Fitzgerald',
 2012: 'John Green',
 1937: 'J.R.R. Tolkien',
 1951: 'J.D. Salinger',
 2000: 'Mary GrandPré',
 1813: 'Jane Austen',
 2003: 'J.K. Rowling',
 2011: 'Veronica Roth',
 1949: 'George Orwell',
 1945: 'George Orwell',
 1947: 'B.M. Mooyaart-Doubleday',
 2009: 'Suzanne Collins',
 1999: 'Mary GrandPré',
 1954: 'J.R.R. Tolkien',
 2010: 'Suzanne Collins',
 2002: 'Alice Sebold',
 1998: 'Mary GrandPré',
 2007: 'Stephenie Meyer',
 1595: 'Robert           Jackson',
 1988: 'Paulo Coelho',
 1993: 'Lois Lowry',
 1950: 'C.S. Lewis',
 1996: 'George R.R. Martin',
 2006: 'Stephenie Meyer',
 1868: 'Louisa May Alcott',
 1847: 'Charlotte Brontë',
 2001: 'Yann Martel',
 1953: 'Ray Bradbury',
 1974: 'Shel Silverstein',
 1979: 'Douglas Adams',
 1932: 'Aldous Huxley',
 1884: 'Guy Cardwell',
 1952: 'E.B. White',
 2015: 'Paula Hawkins',
 1995: 'Philip Pullman',
 2004: 'Jodi Picoult',
 19

E gli associamo il numero di _shared_number_of_reviews_with_a_text:_

In [115]:
for key, value in anno_aut_max.items():
    numero = shared_number_of_reviews_with_a_text_GB[anno_aut_max[key]]
    tupla = anno_aut_max[key]+','+str(numero)
    anno_aut_max[key] = tupla
#anno_aut_max

KeyError: 'Suzanne Collins,340066.0'

In [117]:
anno_aut_max
best_author_of_the_year = pd.DataFrame(anno_aut_max.items())
best_author_of_the_year['Best-Author'] = best_author_of_the_year[1].str.split(expand=True, pat=',')[0]
best_author_of_the_year['Shared-Views-With-Text'] = best_author_of_the_year[1].str.split(expand=True, pat=',')[1]
best_author_of_the_year.drop(labels=[1], axis=1, inplace=True)
best_author_of_the_year.rename({0: 'Year'}, axis='columns', inplace=True)
best_author_of_the_year.sort_values(by='Year', axis=0, ascending=True, inplace=True)
best_author_of_the_year.tail(15)

Year      Best-Author Shared-Views-With-Text
16  1999   Mary GrandPré               136668.0
8   2000   Mary GrandPré               136668.0
30  2001      Yann Martel                42962.0
19  2002     Alice Sebold                36642.0
10  2003     J.K. Rowling               136668.0
39  2004     Jodi Picoult                30719.0
2   2005  Stephenie Meyer               258573.0
27  2006  Stephenie Meyer               258573.0
21  2007  Stephenie Meyer               258573.0
0   2008  Suzanne Collins               340066.0
15  2009  Suzanne Collins               340066.0
18  2010  Suzanne Collins               340066.0
11  2011    Veronica Roth               156896.0
5   2012       John Green               230584.0
37  2015    Paula Hawkins                93600.0

### 9. Assuming that there are no errors in the ISBN fields, find the books in both datasets, and compute the difference of average rating according to the ratings and the goodratings datasets

Abbiamo già calcolato nei punti precedenti l'_average_rating_ per i libri dei dataset _books_ e _good_book,_ perciò utilizziamo per questo calcolo le colonne delle tabelle che si chiamano, rispettivamente per _books_ e _good_book, Average-Rating_ e _average_rating._

In [ ]:
intersection = pd.merge(books, good_book, left_on='ISBN', right_on='isbn') #how=inner è il default
intersection.rename({'Average-Rating': 'Average-Rating-B', 'average_rating': 'Average-Rating-GB'}, axis='columns', inplace=True)
intersection['Difference-Of-Ratings'] = intersection['Average-Rating-B'] - intersection['Average-Rating-GB']
intersection[['ISBN', 'Book-Title', 'Average-Rating-B', 'Average-Rating-GB', 'Difference-Of-Ratings']]

### 10. Split the users dataset according to the age. One dataset contains the users with unknown age, one with age 0-14, one with age 15-24, one with age 25-34, and so on.

Controlliamo se vi sono valori _null_ per il campo _Age_ del dataset _users._

In [ ]:
print(len(users))
len(users[users['Age'].isnull()])

Nel dataset _users,_ su 277348 osservazioni 110204 presentano un valore nullo _(NaN)_ per il campo _Age._

Definiamo una funzione che estrae dal dataset _users_ le osservazioni che hanno un valore di _Age_ compreso tra due valori dati.

In [ ]:
def crea_dataset(a,b):
    dataset=users[(users['Age']>=a)&(users['Age']<=b)]
    return dataset

Per applicarla, verifichiamo quale sia il massimo valore assunto dalla variabile _Age_ nel dataset.

In [ ]:
users['Age'].sort_values(ascending=False).head(15)

Ad un primo impatto si tratta di valori molto strani per quello che rappresentano, tuttavia non abbiamo la sicurezza che l'unità di misura utilizzata sia 'anni', anche se lo ipotizziamo. Verifichiamo quante osservazioni presentino valori più elevati di una simbolica età di 100 anni, per capire se essi possono essere considerati come valori anomali o _outliers_ che si possono eliminare dal dataset.

In [ ]:
len(users[users['Age']>=101])

In [ ]:
sopra_101=len(users[users['Age']>=101])
totali=len(users)

In [ ]:
print(round(sopra_101/totali,3),'%')

Le osservazioni che presentano nel campo _Age_ un valore anomalo, maggiore della soglia simbolica dei 100 anni, sono 364. Poiché rappresentano una piccolissima parte del totale delle osservazioni, lo 0.001%, decidiamo di non considerare nella creazione di dataset per intervalli di 10 anni di età le osservazioni che presentano valori maggiori di 105.



Definiamo dunque i dataset richiesti:

In [ ]:
users_0_14 = crea_dataset(0,14)
users_15_24 = crea_dataset(15,24)
users_25_34 = crea_dataset(25,34)
users_35_44 = crea_dataset(35,44)
users_45_54 = crea_dataset(45,54)
users_55_64 = crea_dataset(55,64)
users_65_74 = crea_dataset(65,74)
users_75_84 = crea_dataset(75,84)
users_85_94 = crea_dataset(85,94)
users_95_104 = crea_dataset(95,104)

E ne stampiamo uno di esempio:

In [ ]:
users_35_44.head()

### 11. Find the books that appear only in the goodbooks datasets.

Effettuiamo una _merge_ dei due dataset con le seguenti opzioni:
- _how = outer,_ che permette di considerare anche gli elementi che si trovano in uno solo dei due dataset che si uniscono;
- _indicator = True,_ che crea una variabile indicante in quale dei due dataset si trovasse l'osservazione _(left_only, right_only, both)._ Questa opzione ci permette successivamente di selezionare le osservazioni che si trovavano solo nel dataset _good_books,_ ovvero quello di destra.

In [ ]:
union = pd.merge(books, good_book, left_on='ISBN', right_on='isbn', how='outer', indicator=True)
union[['ISBN','Book-Title','Book-Author','_merge']].head()

In [ ]:
isbn_only_gb = union[union['_merge'] == 'right_only'][['isbn']]
isbn_only_gb.head()

Una volta ottenuto un dataframe che contiene unicamente gli _isbn_ dei libri che si trovano solo in _good_book,_ effettuiamo un _inner join_ per estrarre e visualizzare anche il titolo e gli autori di ciascun libro.

In [ ]:
books_only_gb = pd.merge(good_book, isbn_only_gb, on='isbn', how='inner')
books_only_gb[['isbn', 'original_title', 'authors']].head()

### 12. Assuming that each pair (author, title) identifies a book, for each book find the number of times it appears in the books dataset. Which books appear the most times?

Per rispondere a questo quesito utilizziamo la funzione _groupby_ rispetto alla coppia _(Book-Author, Book-Title),_ conteggiamo la frequenza delle osservazioni e poi le ordiniamo per conteggio in ordine decrescente, per osservare il libro che appare più frequentemente

In [ ]:
grouped = books.groupby(['Book-Author', 'Book-Title'], as_index=False).count().sort_values(['ISBN'], ascending=False)
grouped.rename({'ISBN': 'Count'}, axis='columns', inplace=True)
grouped[['Book-Author', 'Book-Title', 'Count']].head()

Il libro che appare più frequentemente è _Little Women_ di Louisa May Alcott, per un totale di 21 volte.

### 13. Find the author with the highest average rating according to the goodbooks datasets.

Creiamo un dizionario nel quale al nome di ciascun autore è associata una lista con i diversi ratings che ha ottenuto.

In [ ]:
autori_ratings = dict()
for i, row in good_book.iterrows():
    autori=row['authors'].split(',')
    for autore in autori:
        autore=autore.strip()
      #  autore=strip_multiple_whitespaces(autore)
        if autore not in autori_ratings:
            autori_ratings[autore] = []
        autori_ratings[autore].append(row['average_rating'])
autori_ratings

Successivamente, sostituiamo a ciascuna lista la media dei valori che contiene:

In [ ]:
import statistics

for key, value in autori_ratings.items():
    media = statistics.mean(value)
    autori_ratings[key] = media

In [ ]:
autori_ratings

Ed infine estraiamo l'autore che presenta il rating medio più alto.

In [ ]:
max(autori_ratings, key=autori_ratings.get)

Con un _average_rating_ pari a:

In [ ]:
autori_ratings[max(autori_ratings, key=autori_ratings.get)]
